walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,193 @@
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return freeTime;
+    }
+
+    // To store all the events which the optional attendees are attending.
+    LinkedList<Event> optionalAttendeeEvents = new LinkedList<>();
+
+    /* Changes freeTime such that it now conatins all the free time slots when the meeting can
+    happen, updating the events that optional attendees can attend. */
+    freeSlotsWithMandatoryAttendees(
+        freeTime, events, optionalAttendeeEvents, optionalAttendees, mandatoryAttendees);
+
+    /* If free time is not empty and optional attendees are not empty
+    then go ahead and look for time slots to have meeting with our optional attendees */
+    if (!freeTime.isEmpty() && !optionalAttendeeEvents.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(optionalAttendeeEvents, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  // Loop through all the events given and modify freeTime List
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      LinkedList<Event> optionalAttendeeEvents,
+      Collection<String> optionalAttendees,
+      Collection<String> mandatoryAttendees) {
+
+    for (Event event : events) {
+      /* Check if the event has any optional attendees and
+        save it in the optionalAtttendeeEvents list which we will deal with
+        later in the freeSlotsWithOptionalAttendees method */
+      if (!Collections.disjoint(optionalAttendees, event.getAttendees())) {
+        optionalAttendeeEvents.add(event);
+      }
+      if(!Collections.disjoint(mandatoryAttendees, event.getAttendees())) {
+        findConflictAndResolve(freeTime, event);
+      }
+    }
+
+  }
+
+  // Try checking for any time slots we can get for the meeting with optional attendees
+  private LinkedList<TimeRange> freeSlotsWithOptionalAttendees(
+      LinkedList<Event> optionalAttendeeEvents,
+      LinkedList<TimeRange> freeTime,
+      Collection<String> optionalAttendees) {
+    // Create a new OptionalFreeTime list so that we still have our time slots with mandatory
+    // attendees
+    LinkedList<TimeRange> optionalFreeTime = new LinkedList<>(freeTime);
+    /* Loop through all the events our optional attendees are attending and
+       find free time slots we can have for our meeting */
+    for (Event event : optionalAttendeeEvents) {
+      findConflictAndResolve(optionalFreeTime, event);
+    }
+    /* If we can't find any free time slots with optional attendees
+      then return the original freeTime slots */
+    if (optionalFreeTime.isEmpty()) {
+    return freeTime;
+    }
+
+    return optionalFreeTime;
+  }
```

Comment:
I just read the instructions of this week (I probably should have done this earlier) and handling the optional attendees is an optional challenge. I suggest to remove it from this pull request entirely as it just adds more thing to review. You can skip this challenge or keep it for later as it's more important to get going with the Capstone project so you won't be far behind from the start. If you keep adding new things to the pull request we'll never get to the point where you can submit it.

Another reason to remove this part (for now) is that this code is not quite doing what's required: You should return the meeting slot(s) where the most optional attendees can attend. You only check whether all optional attendees can attend any of the meeting slots.

Reply:
Sorry, I misread the instructions. What you were doing is part of the main project, so please keep it in here.

As we discussed, for future pull requests, please avoid adding features after the the review has started.

Reply:
Sorry and yes I'll make sure to not complicate my pull request.